### PR TITLE
Fix logging issues and move to SyntaxReceiver

### LINKED ISSUE #687: Warning when turning of CsWinRTEnableLogging
**Describe the bug**
If I don't set `CsWinRTEnableLogging` to true in an authoring project, I get the following warning and the winmd isn't generated.

Severity	Code	Description	Project	File	Line	Source	Suppression State	Tool
Warning	CS8785	Generator 'SourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ObjectDisposedException' with message 'Cannot write to a closed TextWriter.'	AuthoringDemo	C:\Users\angzhang\source\repos\AuthoringDemo\AuthoringDemo\CSC	1	Build	Active	Compiler

**Expected behavior**
no warning

**Version Info**
CsWinRT 1.1.1

### PR DESCRIPTION
**Pull request by manodasanW**
Fixing issue with logging triggering ObjectDisposedException when logging is enabled and then turned off.
Moving to using SyntaxReceiver now that we don't rely on visiting all the syntax nodes.

Fixes #687

### CODE REVIEW

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs [hunk 1]
```diff
@@ -31,15 +31,21 @@ public WinRTComponentScanner(GeneratorExecutionContext context, string assemblyN
...
             {
-                var model = _context.Compilation.GetSemanticModel(tree);
-                var nodes = tree.GetRoot().DescendantNodes();
+                var model = _context.Compilation.GetSemanticModel(declaration.SyntaxTree);
 
-                var classes = nodes.OfType<ClassDeclarationSyntax>().Where(IsPublic);
-                foreach (ClassDeclarationSyntax @class in classes)
+                // Check symbol information for whether it is public to properly detect partial types
+                // which can leave out modifier.
+                if(model.GetDeclaredSymbol(declaration).DeclaredAccessibility != Accessibility.Public)
```
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:43]: Does this mean non-public structs will be skipped already? Once you merge I can try those tests I made recently here and check -- might not need my struct PR

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 2]
```diff
@@ -28,91 +31,46 @@ internal sealed class WriteOnlyArrayAttribute : global::System.Attribute
...
         }
 
-        private void GenerateSources(GeneratorExecutionContext context)
+        private void GenerateSources()
         {
-            string cswinrtExe = GetCsWinRTExe(context);
-            string assemblyName = GetAssemblyName(context);
-            string winmdFile = GetWinmdOutputFile(context);
+            string cswinrtExe = context.GetCsWinRTExe();
+            string assemblyName = context.GetAssemblyName();
```
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:69]: Is GetAssemblyName doing the same thing as we were doing? I.e. pulling form analyzerconfigoptions?

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 3]
```diff
@@ -184,27 +138,19 @@ private void GenerateWinMD(MetadataBuilder metadataBuilder, string outputFile)
             peBlob.WriteContentTo(fs);
         }
 
-        private bool CatchWinRTDiagnostics(GeneratorExecutionContext context)
+        private bool CatchWinRTDiagnostics()
```
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:141]: 🚀

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 4]
```diff
@@ -248,9 +196,54 @@ public void Execute(GeneratorExecutionContext context)
...
+                syntaxNode is NamespaceDeclarationSyntax)
+            {
+                Declarations.Add(decaralation);
+            }
+        }
+
+        private bool IsPublic(MemberDeclarationSyntax member)
+        {
+            // We detect whether partial types are public using symbol information later.
+            return member.Modifiers.Any(m => m.IsKind(SyntaxKind.PublicKeyword) || m.IsKind(SyntaxKind.PartialKeyword));
```
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:246]: can you have a `internal partial class MyClass` ?

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 5]
```diff
@@ -28,91 +31,46 @@ internal sealed class WriteOnlyArrayAttribute : global::System.Attribute
...
-        }
-
-        private static bool IsCsWinRTComponent(GeneratorExecutionContext context)
-        {
-            if (context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.CsWinRTComponent", out var isCsWinRTComponentStr))
-            {
-                return bool.TryParse(isCsWinRTComponentStr, out var isCsWinRTComponent) && isCsWinRTComponent;
-            }
+        private Logger Logger { get; }
+        private readonly GeneratorExecutionContext context;
```
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:35]: We can still report diagnostics on a `readonly` context?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs (hunk 1 above)
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:43]: Also add a space between `if` and `(model.GetDeclaredSymbol`

**Comment by j0shuams** [comment]
Looks good, approved with assumption DiagnosticTests succeed

**Review by j0shuams — APPROVED** [review]

File: src/Authoring/WinRT.SourceGenerator/Generator.cs (hunk 4 above)
Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/Generator.cs:246]: Yes you can which will be caught by the DeclaredAccessibility check done later on using the symbol information.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs (hunk 1 above)
Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:43]: Yes non public structs should be skipped.

File: src/Authoring/WinRT.SourceGenerator/Generator.cs (hunk 5 above)
Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/Generator.cs:35]: The value we got was always readonly so we should still be able to report diagnostics without issues.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs (hunk 1 above)
Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs]: Also did a format document on this and the other files.

**Pull request merged**